ndhistogram/src/axis/cyclic.rs

```diff
@@ -0,0 +1,199 @@
+use super::{Axis, BinInterval, Uniform};
+use std::fmt::Debug; // TODO Display
+
+use num_traits::{Float, Num, NumCast, NumOps};
+use serde::{Deserialize, Serialize};
+
+/// A wrap-around axis with equal-sized bins.
+///
+/// An axis with `N` equally-spaced, equal-sized bins, in `[low, high)`.
+/// Entries outside this interval get wrapped around.
+/// There are no overflow bins so this axis has exactly `N` bins.
+///
+/// # Examples
+/// 1D histogram with 4 bins distributed around a circle.
+/// ```
+/// use ndhistogram::{ndhistogram, Histogram};
+/// use ndhistogram::axis::{Axis, BinInterval, Cyclic};
+/// let mut hist = ndhistogram!(Cyclic::new(4, 0.0, 360.0));
+/// hist.fill(& 45.0         ); // Add entry at 45 degrees
+/// hist.fill(&(45.0 + 360.0)); // Add entry at 45 degrees + one whole turn
+/// hist.fill(&(45.0 - 360.0)); // Add entry at 45 degrees + one whole turn backwards
+/// // All 3 above entries end up in the same bin
+/// assert_eq!(hist.value(&45.0), Some(&3.0));
+/// // Lookup also wraps around
+/// assert_eq!(hist.value(&(45.0 + 360.0)), Some(&3.0));
+/// assert_eq!(hist.value(&(45.0 - 360.0)), Some(&3.0));
+/// ```
+/// Time of day
+/// ```
+/// use ndhistogram::{ndhistogram, Histogram};
+/// use ndhistogram::axis::{Axis, BinInterval, Cyclic};
+/// let bins_per_day = 24;
+/// let hours_per_bin = 1;
+/// let start_at_zero = 0;
+/// let four_pm = 16;
+/// let mut hist = ndhistogram!(Cyclic::with_step_size(
+///     bins_per_day, start_at_zero, hours_per_bin
+/// ));
+/// hist.fill(&40);                               // The 40th hour of the week ...
+/// assert_eq!(hist.value(&four_pm), Some(&1.0)); // ... is at 4 pm.
+/// ````
+#[derive(Default, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Serialize, Deserialize)]
+pub struct Cyclic<T = f64> {
```

Comment:
I guess that you also plan to create a `Cyclic` axis where internal axis is `Variable`?

In that case I guess that it would be better to name this struct `UniformCyclic` (and then the `Variable` axis would be named `VariableCyclic`).

Reply:
Hadn't really given it any thought (I haven't found the need for it myself). Or if I had, I've forgotten about it by now (it's been a while since I wrote this code).

I've changed the names, in anticipation of implementing `VariableCyclic` ... although I've changed it to the wrong name, so that needs fixing first!

Reply:
I've pushed a first stab at `VariableCyclic`.

Still haven't written the main doc comments. Do you have any ideas for a good example to show in the doctests?

Reply:
This code duplication 
+ https://github.com/jacg/ndhistogram/blob/cyclic-axis/ndhistogram/src/axis/variablecyclic.rs#L56-L66
+ https://github.com/jacg/ndhistogram/blob/cyclic-axis/ndhistogram/src/axis/uniformcyclic.rs#L95-L104
is annoying.
Unfortunately, the code relies on `self.axis` which is just an implementation detail of these two types; so generics won't work. A macro maybe?

Reply:
Yes, sure, generics would work if another trait is introduced, but I don't think it's worth it. The implementations (of `index`) are identical, and the `macro_rules!` macro definitely works (I've tried it), but you'd have to find somewhere to put it, which is also probably more trouble than it's worth.

So, I'd be inclined to live with the duplication.

Reply:
Ok, great. It sounds like we agree. Let's stick with the current implementation.